docs/tutorials-redeliver-messages.md

```diff
@@ -0,0 +1,236 @@
+4. data loss exists in the case of TCP long connections, although the probability is extremely low.
+   ![img_8.png](../static/img/blog-consume-best-practice/Ack-hole.png)
+
+For case 1, configuring `deadLetterPolicy` is a good solution. When the business system receives a signal of message
+processing failure, it can immediately check and decide whether to retry after a period of time. After they call
+`reconsumeLater` API, the message will be acknowledged and resent to the retry letter topic that is automatically
+subscribed by the consumer. When it reaches `maxRedeliverCount`, the message will be sent to the dead letter topic.
+Messages sent to the dead letter topic should be considered as non-retryable messages, and we recommend setting an
+initialSubscriptionName to avoid being deleted by the retention policy and then, let maintenance personnel regularly
+handle non-retryable messages in the dead letter topic.
+![img_2.png](../static/img/blog-consume-best-practice/DLQ.png)
+```java
+        Consumer<Integer> consumer = pulsarClient.newConsumer(Schema.INT32)
+        .deadLetterPolicy(DeadLetterPolicy.builder()
+                .maxRedeliverCount(maxRedeliverCount)
+                .deadLetterTopic(deadLetterTopic)
+                .retryLetterTopic(retryLetterTopic)
+                .initialSubscriptionName(initialSubscriptionName)
+                .build())
+        .subscriptionType(SubscriptionType.Shared)
+        .enableRetry(true)
+        .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+        .topic(topic)
+        .subscriptionName(sub)
+        .subscribe();
+Message<Integer> message = consumer.receive();
+        try{
+            // Process message
+            consumer.acknowledge(message);
+        }catch(Exception e){
+            // Check whether to redeliver the message again.
+            if(e instanceof RetryException){
+                consumer.reconsumeLater(message, delay, timeunit);
+            }
+            consumer.reconsumeLater(message, delay, timeunit);
+        }
+```
+
+For case 2, where the business system gets stuck or an error occurs which causes the received message to be missed for
+processing, configuring ack timeout is a good solution. The consumer will record every message received on the client side.
+If these messages have not been acknowledged after the specified time, the consumer will request the broker to resend these
+messages to other brokers.
+**Suggestion:** The ack timeout should be set slightly longer based on the message processing speed of the business system.
+If the business system is still processing messages, but the processing time is too long or the timeout is set too small,
+it may result in duplicate message consumption.
+![img_3.png](../static/img/blog-consume-best-practice/AckTimeout.png)
+```java
+        Consumer<Integer> consumer = pulsarClient.newConsumer(Schema.INT32)
```

Comment:
Code indentation probably should be fixed here and in the other code blocks. 

<img width="675" alt="Screenshot 2024-04-02 at 9 40 49 AM" src="https://github.com/apache/pulsar-site/assets/9302460/a2cf06bc-1adf-45b1-95da-e1c3b6b6370a">

Reply:
Fixed. Thanks for the suggestion.